### PR TITLE
Change user search to partially match based on user name instead of exact matching

### PR DESCRIPTION
**Pull request by ishchow**
I made some changes to the user search fragment in order to allow the users to searched without exact matching. So if I typed "ish" in the fragment, I should receive "ishaatchowdhury@outlook.com", "ishaat@ualberta.ca", and "ishaatchowdhury@gmail.com" instead of nothing. In addition, the search results change as the user types which makes the search more intuitive for the user.

In addition, there are several related changes as part of this PR in order to improve maintainability and performance of search:
1. Modify FirebaseQueryLiveData to allow the query to be dynamically changed at runtime. This is what allows the as you type search to take place.
2. Move the Firebase querying logic to a ViewModel so that this information isn't contained in the view, making the view easier to understand and change.
3. Add a generic ListAdapter for a list of Users to interact with a RecyclerView. When the underlying list of the adapter changes by a call to submitList(), the new and old lists are diffed in a background thread improving UI responsiveness. This is provided by a Google library.
4. Use a RecyclerView to render the list of usernames instead of a ListView. Previously, a new list of usernames had to be created in order to properly render the list when using a ListView. This processing is removed by using the RecyclerView since the username of the User can be bound to the item of the userList without generating a new list of usernames to pass into an ArrayAdapter. This is an improvement in terms of speed and maintainability since RecyclerViews are in general much more efficient than ListViews.

### CODE REVIEW

**Comment by ishchow** [comment]
> 
> 
> I tested this on my device. This is a great change in terms of functionality and architecture!
> 
> The only minor change I can maybe recommend is to remove the initialization of LiveData of userListLiveData in SearchFragment.java since we can just observe directly from the LiveData in the ViewModel.
> 
> Otherwise, I think we can merge it.

Thanks for the review Marcus!

I'm not really sure what you mean by this exactly though. I expose the livedata in the search fragment and observe for changes on it there so that the list adapter can update the contents of it's list based on the new data. If I were to add the adapter to the view model and observe for changes in the live data within the view model then that would make the view model be less flexible since it requires that the adapter be composed in the view model. Unless of course I'm misunderstanding your comment.

**Comment by ishchow** [comment]
As part of this change, I also added some indexes in the database to make the queries even faster.

fb_new_rules

File: app/src/main/java/com/example/atheneum/fragments/SearchFragment.java [hunk 1]
```diff
@@ -82,80 +56,41 @@
...
+                    ((MainActivity)getActivity()).passDataToViewProfileActivity(user);
                 }
-                defaultUserNameList = userNameList;
-                ArrayAdapter<String> adapter = new ArrayAdapter<>(getActivity(), android.R.layout.simple_list_item_1, userNameList);
-                userListView.setAdapter(adapter);
             }
+        });
 
+        searchUsersViewModel = ViewModelProviders.of(getActivity()).get(SearchUsersViewModel.class);
+        LiveData<List<User>> userListLiveData = searchUsersViewModel.getUserListLiveData();
```
Comment (MarcusBoay) [inline app/src/main/java/com/example/atheneum/fragments/SearchFragment.java]: I was mentioning this in my previous comment. The google repo (https://github.com/googlesamples/android-architecture-components/blob/master/BasicSample/app/src/main/java/com/example/android/persistence/ui/ProductFragment.java) for their UI on lines 85 and 93 show them calling a method to get the LiveData and calling the .observe() method immediately. I tested removing this and simply replacing line 83 and 84 with the code below and it seemed to work fine.
```
searchUsersViewModel.getUserListLiveData().observe(...);
```

Comment (ishchow) [reply app/src/main/java/com/example/atheneum/fragments/SearchFragment.java]: Fixed

**Pull request merged**